libvast/src/plugin.cpp

```diff
@@ -246,17 +246,28 @@ load(const std::vector<std::string>& bundled_plugins,
 
 /// Initialize loaded plugins.
 caf::error initialize(caf::actor_system_config& cfg) {
+  auto global_config = record{};
+  auto global_opts = caf::content(cfg);
+  if (auto global_opts_data = to<record>(global_opts)) {
+    global_config = std::move(*global_opts_data);
+  } else {
+    VAST_DEBUG("unable to read global configuration options: {}",
+               global_opts_data.error());
+  }
+  auto plugins_record = record{};
+  if (global_config.contains("plugins")) {
+    if (auto plugins_entry
+        = caf::get_if<record>(&global_config.at("plugins"))) {
+      plugins_record = std::move(*plugins_entry);
+    }
+  }
+  VAST_DEBUG("collected {} global options for plugin initialization",
+             global_config.size());
   for (auto& plugin : get_mutable()) {
     auto merged_config = record{};
-    // First, try to read the configuration from the merged VAST configuration.
-    if (auto opts = caf::get_if<caf::settings>(
-          &cfg, fmt::format("plugins.{}", plugin->name()))) {
-      if (auto opts_data = to<record>(*opts))
-        merged_config = std::move(*opts_data);
-      else
-        VAST_DEBUG("unable to read plugin options from VAST configuration at "
-                   "plugins.{}: {}",
-                   opts_data.error(), plugin->name());
+    // First, try to read the configurations from the merged VAST configuration.
+    if (plugins_record.contains(plugin->name())) {
+      merged_config = caf::get<record>(plugins_record.at(plugin->name()));
```

Comment:
This crashes for a config file like this:

```
plugins:
  myplugin: foo
```

Because the code here assumes that `plugins.myplugin` holds a record, when in this example it holds a string. The solution would be to use `caf::get_if` here instead to verify that it holds a record. It's probably a hard error if it doesn't.

Reply:
I added a validity check in https://github.com/tenzir/vast/pull/2949/commits/0a7f27d4225433ada4bb26c9b2ddab89abd5d609